### ISSUE TITLE
AS no longer works

### ISSUE THREAD

**Issue by a-pirard** [JOSM/areaselector#25]
Since an unknown time, Area Selector no longer works.  It draws 4 ways at the border of the window.
Neither with BE SPW PICC, nor with BE AGIV I was told, and not even as I can see with AT basemat.at which is your testbed I think.
I don't think the problem occurred with an AS update but rather with a JOSM one.
I ran josm-snapshot-10693.jar and AS did run, but very badly again.
It does not draws its ways onto the building walls but around them.
However, that shows better some interested friends what they will have when it will run superbly again.
TIA for fixing this !!!

**Comment by r00tat** [comment]
Sounds like the layers are not rendered correctly. I'll try to look into it this weekend!

**Comment by r00tat** [comment]
#24 might also be related

**Comment by a-pirard** [comment]
Would it help you if I dichotomically tried other JOSM versions and found the last working one and hence the version introducing incompatibility? It would be a workaround, too.

**Comment by a-pirard** [comment]
I replied by e-mail but it didn't go through. So, here's a copy&paste.

Hi Paul,

I made those tests and I determined that JOSM started drawing on the window border beginning with josm-snapshot-10819.jar.
But before that, ending with 10799, the way it draws is something oval around and far from the walls.
So, I didn't find the workaround I was looking for.
The complete list of my tests is below.
I hope it will help you somehow.

Cheers

André.

> 10693 oval
> 10753 oval
> 10786 oval
> 10797 oval
> 10799 oval
> 10819 border
> 10929 border
> 11163 border

**Comment by a-pirard** [comment]
On 2016-10-25 17:52, a-pirard wrote:
>
> Would it help you if I dichotomically tried other JOSM versions and
> found the last working one and hence the version introducing
> incompatibility? It would be a workaoval, too.
>

Hi Paul,

I made those tests and I determined that JOSM started drawing on the
window border beginning with josm-snapshot-10819.jar
<https://josm.openstreetmap.de/download/josm-snapshot-10819.jar>.
But before that, ending with 10799, the way it draws is something oval
around and far from the walls.
So, I didn't find the workaround I was looking for.
The complete list of my tests is below.
I hope it will help you somehow.

Cheers

André.


> 10693 oval
> 10753 oval
> 10786 oval
> 10797 oval
> 10799 oval
> 10819 border
> 10929 border
> 11163 border

**Comment by r00tat** [comment]
I did some debugging and found that the layers are not painted to the image.

This is the line, which no longer works:
[AreaSelectorAction Line 160](https://github.com/JOSM/areaselector/blob/master/src/org/openstreetmap/josm/plugins/areaselector/AreaSelectorAction.java#L160)
```
layer.paint(imgGraphics, mapView, mapView.getRealBounds());
```

I'll try to find out what has changed in JOSM core with the help of the revisions you mentioned.

**Comment by r00tat** [comment]
looks like the AbstractTileSourceLayer does not draw any more via paint:

```
    @Override
    public void paint(Graphics2D g, MapView mv, Bounds bounds) {
        // old and unused.
    }
```

Drawing is now done via doPaint and drawInView in AbstractTileSourceLayer

```
        private void doPaint(MapViewGraphics graphics) {
            drawInViewArea(graphics.getDefaultGraphics(), graphics.getMapView(), graphics.getClipBounds().getProjectionBounds());
        }
```

The problem is, that drawInViewArea is private, so this method cannot be used to get the imagery. 

@don-vip I saw you made the changes to [AbstractTileSourceLayer](https://josm.openstreetmap.de/browser/josm/trunk/src/org/openstreetmap/josm/gui/layer/AbstractTileSourceLayer.java?rev=10757). As painted is deprecated and drawInViewArea is deprecated, how can the imagery be accessed via plugin?

**Comment by a-pirard** [comment]
Thanks for working on this, Paul !!!

As we are deep in the subject, may I please mention #20 which I meant not as a question but as a suggestion.  If I'm correct: Distance tolerance setting should be in cm.
Because different zooms produce different results, I think that you are working on the zoomed image, not on the server's image.
Hence, the tolerance dots numbers must be zoomed too (multiplied by the level) and you must say to which zoom the setting applies.
Or the setting should be in cm that need no explanation and that you must convert to the correct dot numbers at each zoom level.
Duplicating this comment to #20.

**Comment by don-vip** [comment]
@michaelzangl it seems I have broken AreaSelector plugin with core compatibility changes, do you see how to fix it?

**Comment by r00tat** [comment]
@don-vip @michaelzangl the easiest solution would be to call drawInViewArea in paint. I don't know if that has any side effects.

**Comment by r00tat** [comment]
I tested reimplementing paint and it looks like this will work fine.

This is my fix: 
```
    @Override
    public void paint(Graphics2D g, MapView mv, Bounds bounds) {
        // old and unused.
        drawInViewArea(g, mv, mv.getProjectionBounds());
    }
```

**Comment by michaelzangl** [comment]
This won't help much. As soon as #13386 is implemented, it won't work any more ;-).

The changes were intended to make JOSM support multiple map views. All layers should use the `attachToMapView(MapViewPaintable.MapViewEvent event)` method. It receives the map view instance and is called as soon as the map view is created. `paint` is still called by the map view if no other implementation is present.

It should return a `MapViewPaintable.LayerPainter`. You can move methods like `getLayeredImage`, ... there. That way, you do not need any global Main.map.mapView references. Nice side effect: If you do not keep any references to it (listeners, ...) it gets garbage collected as soon as the map view is destroyed.

There is currently no way to access the paintes that are used by the current map view.

You would need to access the `registeredLayers` field. The best way would be to make the method `paintLayer` accessible. I'd suggest creating a new method. Bounds can be computed automatically.
{{{
public void paintLayers(Iterable<? extends Layer> layer, Graphics2D g);
}}}
Would this be enough for you?

Some more notes while looking at your code:
`getLayeredImage`: You can use `getVisibleLayersInZOrder()`, then only filter for background layers.

`createArea()`: You can use the JOSM preference system for this. We use '.' to separate child preferences, but this is just a convention. You can use Boolean/Integer/String/... preferences.
{{{
KEY_TOLERANCEDIST = "areaselector.tolerancedist";
TOLERANCEDIST = new DoubleProperty(KEY_TOLERANCEDIST, DEFAULT_TOLERANCEDIST);
...
distMeters = TOLERANCEDIST.get(); // no try/catch, no if/else.
}}}

**Comment by r00tat** [comment]
Once there are multiple mapViews possible, there should be a way to access the active mapView from a plugin. I suppose there is an idea how to handle that, right?

I believe it would be better to provide a method to paint one layer, not all together in one method. A plugin might want to access only specific layers and therefore the existing paint function would is easy to use. 


Regarding getLayeredImage:
Thanks, Done. (d57f5881ef9cf56db4a6cb84a003e14a1ee58658)

Regarding the preferences:
Thank's for your suggestion, I'll change it. It makes the code way more readable.

**Comment by r00tat** [comment]
I just looked at `paintLayer`in MapView and providing access to that method would help. 
Providing a method to paint multiple layers at once or not is more a matter of code style or personal taste.

**Comment by michaelzangl** [comment]
There you go:

https://josm.openstreetmap.de/changeset/11226/josm

It is best to feed it with the layers obtained using: `mapView.getLayerManager().getVisibleLayersInZOrder()`

Let me know if there are any issues with using the method.

**Comment by r00tat** [comment]
Thanks @michaelzangl for implementing the new method in JOSM core!

**Comment by a-pirard** [comment]
On 2016-11-10 19:31, Paul wrote:
> Closed #25 <https://github.com/JOSM/areaselector/issues/25> via
> ae81ffe
> <https://github.com/JOSM/areaselector/commit/ae81ffea4510e01a9f92f197073c86c11fee8087>.
> Thanks @michaelzangl <https://github.com/michaelzangl> for
> implementing the new method in JOSM core!
You JOSM people are great people, really.  THANKS !!!!
I wish everybody doing serious work were using it.

Are we supposed to use (test) it right now?
I ran josm-snapshot-11227, updated the plugins and it contoured the
whole window again.

Additionally, when activating my layer, I get the following dialog:
> The layer BE SPW PICC 0 does not support the new projection EPSG:3857.
> Supported projections are:
> Change the projection again or remove the layer.
(copy&paste not invented yet, middle-age copying used)

"OK" and the layer seems to display all right, though. It's definition:
> wms:http://geoservices.wallonie.be/arcgis/rest/services/TOPOGRAPHIE/PICC_VDIFF/MapServer/export?&f=image&format=png8&dpi=90&transparent=True&SRS={proj}&bboxSR={wkid}&imageSR={wkid}&bbox={bbox}&size={width},{height}&layers=hide:0,12,26
THANKS !!!!

Cheers
Cordialement,

André.

**Comment by r00tat** [comment]
I just released the new plugin version, it should be available in the plugin section in a few minutes. 

The minimum JOSM version is now 11226. 

The projection error message is something from an update from JOSM core, not by the plugin.
